src/miral/basic_window_manager.cpp

```diff
@@ -116,13 +116,19 @@ auto miral::BasicWindowManager::add_surface(
 
     WindowSpecification const& spec = policy->place_new_window(session_info, place_new_surface(session_info, params));
     scene::SurfaceCreationParameters parameters;
+    if (params.depth_layer.is_set())
+        parameters.depth_layer = params.depth_layer;
     spec.update(parameters);
     auto const surface_id = build(session, parameters);
-    Window const window{session, session->surface(surface_id)};
+    auto surface = session->surface(surface_id);
+    Window const window{session, surface};
     auto& window_info = this->window_info.emplace(window, WindowInfo{window, spec}).first->second;
 
     if (spec.parent().is_set() && spec.parent().value().lock())
+    {
         window_info.parent(info_for(spec.parent().value()).window());
+        surface->set_depth_layer(spec.parent().value().lock()->depth_layer());
```

Comment:
I'm uncomfortable with this

    if (... && foo.lock())
    {
        bar(..., foo.lock()->dereference_here())
    }

setup. We've locked the WindowManager, so we know that the `weak_ptr<Surface>→ WindowInfo&` lookup should succeed, but do we know that the `weak_ptr<Surface>` is still valid here? It would be simple to capture the `shared_ptr<Surface>` from the conditional to ensure that the pointer is still valid.

Reply:
I got a funny feeling when I wrote this, now I see why.